ESP32-C3: Add support for TWAI / CAN

martinjaeger opened the pull request:
This PR updates the existing Espressif TWAI driver (based on an SJA1000 backend) to support the newer ESP32-series MCUs like ESP32-C3 and ESP32-S2, which have slightly incompatible registers compared to the original ESP32 and the SJA1000 controller.

## Background

Based on my comparison of the datasheets and the ESP-IDF implementation the following registers are affected:

- `TWAI_BUS_TIMING_0_REG` (`CAN_SJA1000_BTR0`)
    - `TWAI_BAUD_PRESC` has size of 13 bits instead of 6 bits
    - `TWAI_SYNC_JUMP_WIDTH` is moved to BIT(14) instead of BIT(6)

- `TWAI_CLOCK_DIVIDER_REG` (`CAN_SJA1000_CDR`)
    - `TWAI_CD` has length of 8 bits instead of 3 bits
    - `TWAI_CLOCK_OFF` at bit 8 instead of bit 3
    - `TWAI_EXT_MODE` bit missing (always "extended" = PeliCAN mode)

- `TWAI_STATUS_REG`
    - new bit 8: `TWAI_MISS_ST` This bit reflects whether the data packet in the RX FIFO is complete. 1: The current
packet is missing; 0: The current packet is complete (RO)

- `TWAI_INT_RAW_REG`
    - new bit 8: `TWAI_BUS_STATE_INT_ST`

- `TWAI_INT_ENA_REG`
    - new bit 8: `TWAI_BUS_STATE_INT_ENA`

See also discussions in previous PR: https://github.com/zephyrproject-rtos/zephyr/pull/48124#discussion_r926961797

This PR considers the changes in the clock divider and bus timing registers, but does not make use of the new missed frame flag and state interrupt.

## Testing

The `esp32c3_devkitm` board can be used for testing without CAN transceiver when CAN TX and RX pins (GPIO4 and GPIO5 by default) are connected with a jumper.

As twister had some issues with flashing and testing via the same serial I ran the tests in `tests/drivers/can` and `tests/subsys/canbus/isotp` manually. They are all passing.

martinjaeger committed:
drivers: can: sja1000: use generic can_set_timing API

This is a preparation to use the custom implementation in ESP32 TWAI
frontend for newer ESP32 series chips like the ESP32-C3, which have
timing registers incompatible with the SJA1000.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
soc: riscv: esp32c3: include espressif's soc.h

The soc.h from the espressif HAL module provides register definitions
like APB_CLK_FREQ, which are required for the Zephyr TWAI driver.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
drivers: can: esp32_twai: add support for newer MCUs

Newer ESP32 series MCUs like the ESP32-C3 contain some register changes
incompatible to the original ESP32 and the SJA1000.

The additions in this commit consider these changes and fix the
incompatibilities in the TWAI front-end for the SJA1000 driver.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
dts: riscv: espressif: esp32c3: add TWAI node

Add nodes to enable the TWAI (CAN bus) peripheral.

Signed-off-by: Martin Jäger <martin@libre.solar>

martinjaeger committed:
boards: riscv: esp32c3_devkitm: enable TWAI peripheral

The board does not have a CAN transceiver on board, but the driver
can be tested in loopback mode with the CAN RX and TX pins shorted.

Signed-off-by: Martin Jäger <martin@libre.solar>

henrikbrixandersen commented on drivers/can/can_esp32_twai.c:
```diff
@@ -54,6 +64,49 @@ static int can_esp32_twai_get_core_clock(const struct device *dev, uint32_t *rat
 	return 0;
 }
 
+static int can_esp32_twai_set_timing(const struct device *dev, const struct can_timing *timing)
```
I would suggest moving the comments and definitions from the internal header file to this file to improve readability, only define the actually needed register bit masks (avoid defining aliases for the already existing registers as this makes it much more difficult to spot the actual differences.

Encapsulate the whole thing in `#ifdef` as none of this is needed for the original ESP32. No need to define unused aliases for the ESP32 as no code uses them.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+ */
+#define TWAI_BUS_TIMING_0_REG           CAN_SJA1000_BTR0
+
+#ifdef CONFIG_SOC_ESP32
+#define TWAI_BAUD_PRESC_MASK            CAN_SJA1000_BTR0_BRP_MASK
+#define TWAI_SYNC_JUMP_WIDTH_MASK       CAN_SJA1000_BTR0_SJW_MASK
+#else
```
No need for these definitions.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+ *   - TWAI_BAUD_PRESC has size of 13 bits instead of 6 bits
+ *   - TWAI_SYNC_JUMP_WIDTH is moved to BIT(14) instead of BIT(6)
+ */
+#define TWAI_BUS_TIMING_0_REG           CAN_SJA1000_BTR0
```
No need for aliases.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+
+#include "can_sja1000_priv.h"
+
+/*
+ * TWAI_BUS_TIMING_0_REG is incompatible to CAN_SJA1000_BTR0:
+ *   - TWAI_BAUD_PRESC has size of 13 bits instead of 6 bits
+ *   - TWAI_SYNC_JUMP_WIDTH is moved to BIT(14) instead of BIT(6)
+ */
```
Move this comment to the custom `set_timing()` function to document why it is needed.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+#define TWAI_BAUD_PRESC_PREP(brp)	FIELD_PREP(TWAI_BAUD_PRESC_MASK, brp)
+#define TWAI_SYNC_JUMP_WIDTH_PREP(sjw)	FIELD_PREP(TWAI_SYNC_JUMP_WIDTH_MASK, sjw)
+
+/*
+ * TWAI_BUS_TIMING_1_REG is compatible to CAN_SJA1000_BTR1 and only defined for consistency
+ */
+#define TWAI_BUS_TIMING_1_REG           CAN_SJA1000_BTR1
+
+#define TWAI_TIME_SEG1_MASK             CAN_SJA1000_BTR1_TSEG1_MASK
+#define TWAI_TIME_SEG2_MASK             CAN_SJA1000_BTR1_TSEG2_MASK
+#define TWAI_TIME_SAMP                  CAN_SJA1000_BTR1_SAM
+
+#define TWAI_TIME_SEG1_PREP             CAN_SJA1000_BTR1_TSEG1_PREP
+#define TWAI_TIME_SEG2_PREP             CAN_SJA1000_BTR1_TSEG2_PREP
```
While it may be more consistent it also makes the code more convoluted. Just use the original SJA1000 definitions where there are no changes.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+#define TWAI_TIME_SEG1_PREP             CAN_SJA1000_BTR1_TSEG1_PREP
+#define TWAI_TIME_SEG2_PREP             CAN_SJA1000_BTR1_TSEG2_PREP
+
+/*
+ * TWAI_CLOCK_DIVIDER_REG is incompatible with CAN_SJA1000_CDR
+ *   - TWAI_CD has length of 8 bits instead of 3 bits
+ *   - TWAI_CLOCK_OFF at BIT(8) instead of BIT(3)
+ *   - TWAI_EXT_MODE bit missing (always "extended" = PeliCAN mode)
+ */
```
Move this to the init function to document why the `#ifdef` is needed.

henrikbrixandersen commented on drivers/can/can_esp32_twai_priv.h:
```diff
@@ -0,0 +1,71 @@
+ *   - TWAI_CLOCK_OFF at BIT(8) instead of BIT(3)
+ *   - TWAI_EXT_MODE bit missing (always "extended" = PeliCAN mode)
+ */
+#define TWAI_CLOCK_DIVIDER_REG          CAN_SJA1000_CDR
```
No need for an alias.

henrikbrixandersen reviewed (changes requested):
Looks good, but I would prefer less aliases and moving the comments closer to the code using them. This would make the introduced header file almost empty and the remaining definitions can go directly into the C file.

martinjaeger commented on boards/riscv/esp32c3_devkitm/esp32c3_devkitm.yaml:
```diff
@@ -4,6 +4,8 @@ type: mcu
 toolchain:
   - zephyr
+supported:
+  - can
```
I added it to run twister with `-t can` but it didn't work anyways. Removed.

martinjaeger commented:
Thanks @henrikbrixandersen for your quick review. I agree with your suggestions and updated the commits. Please have a look if it meets your expectations.

henrikbrixandersen commented on drivers/can/can_esp32_twai.c:
```diff
@@ -90,6 +179,18 @@ static int can_esp32_twai_init(const struct device *dev)
+	 *
+	 * Resetting the register to 0 to remove the the mode bit set in can_sja_1000_init().
+	 */
+	can_esp32_twai_write_reg32(dev, CAN_SJA1000_CDR, 0U);
```
Hardcoding this to `0U` removes the possibility of turning off the clock output or setting the divisor?

martinjaeger replied:
As far as I understand this register is useless for ESP32, so I reset the clock divider to zero (as done in the SJA1000 backend as well).

The *external* CLKOUT configured with this register could be used for an external MCU in case of an actual SJA1000 chip. But it doesn't really seem to make sense if the SJA1000 is built directly *into* the SOC.

The CLKOUT is also more or less undocumented in the ESP32-C3 reference manual. If you search for it, the only hit (except for the register definition itself) is a table which says that the signal is not muxed to any pin.

henrikbrixandersen replied:
The current TWAI driver sets the CDR register based on the `clkout-divider` DTS property obtained via the `CAN_ESP32_TWAI_DT_CDR_INST_GET()` macro. (see https://github.com/zephyrproject-rtos/zephyr/blob/98580f1c22481c200db2f37881e9c5161f2bd35f/dts/bindings/can/espressif%2Cesp32-twai.yaml#L26 and https://github.com/zephyrproject-rtos/zephyr/blob/98580f1c22481c200db2f37881e9c5161f2bd35f/drivers/can/can_esp32_twai.c#L119).

Pinctrl for muxing the CLKOUT signal for ESP32* was added in https://github.com/zephyrproject-rtos/hal_espressif/pull/108 and fully supported.

While using this signal may or may not make sense for your use-case, I would like to retain support for this no matter if the driver is used on the original ESP32 or a later variant.

martinjaeger replied:
Interesting. I still can't find this feature being documented anywhere in the ESP32 and ESP32-C3 reference manual.

Digging deeper into the Espressif HAL reveals that the divider can now be 1 or an even number from 2 to 490 (!):

https://github.com/zephyrproject-rtos/hal_espressif/blob/68e62d9989e8a8b250348ef28aca1c70da00b0c1/components/hal/esp32c3/include/hal/twai_ll.h#L643-L665

I think your previous approach with a dedicated `#define` to translate divider to register doesn't make sense anymore, as I don't want to create 490/2 defines. Also the enum in the binding would get very long...
Any suggestion how to fix this in a clean way? Remove the enum restricting the values in the binding and do a check + calculation similar to the Espressif HAL in `can_esp32_twai_init`?

henrikbrixandersen replied:
Agreed. Using `BUILD_ASSERT` sounds like a good solution.

henrikbrixandersen replied:
Remember to update the binding documentation as well.

martinjaeger replied:
The macros are getting a bit convoluted, but it's finally still possible to evaluate the `clkout-divider` at compile-time. Please have a look again. Binding doc is also updated.

henrikbrixandersen reviewed (changes requested):
Thanks. I think that's much more readable. Just one last comment/question:

henrikbrixandersen commented on drivers/can/can_esp32_twai.c:
```diff
@@ -18,6 +19,31 @@
+ */
+#ifndef CONFIG_SOC_ESP32
+
+/* TWAI_BUS_TIMING_0_REG is incompatible with CAN_SJA1000_BTR0 */
+#define TWAI_BAUD_PRESC_MASK            GENMASK(12, 0)
+#define TWAI_SYNC_JUMP_WIDTH_MASK       GENMASK(15, 14)
+#define TWAI_BAUD_PRESC_PREP(brp)	FIELD_PREP(TWAI_BAUD_PRESC_MASK, brp)
+#define TWAI_SYNC_JUMP_WIDTH_PREP(sjw)	FIELD_PREP(TWAI_SYNC_JUMP_WIDTH_MASK, sjw)
```
Doesn't this call for new timing_max values as well?

martinjaeger replied:
Good point. I saw that you had to exclude tests with very low baudrate. Maybe supporting low baud rates with high clock speed was the reason why they extended the prescaler register width.

I'll provide a fix for that as well.

martinjaeger replied:
Done. Unfortunately leads to two more `#ifdef`s.

henrikbrixandersen commented on drivers/can/can_esp32_twai.c:
```diff
@@ -90,17 +179,44 @@ static int can_esp32_twai_init(const struct device *dev)
+		.prop_seg = 0,						\
+		.phase_seg1 = 16,					\
+		.phase_seg2 = 8,					\
+		.prescaler = (1U << 13)					\
```
```suggestion
		.prescaler = BIT(13)						\
```

You could skip the definition and just assign these directly to `timing_max` with a comment - this would reduce the number of `#ifdef`s.

martinjaeger replied:
Done. And switched to hex numbers, as they also allow to easily spot the register size (in contrast to decimal numbers). I prefer using the `BIT(x)` macro only for actual bit flags, but not to specify the maximum size of a number.

martinjaeger commented on drivers/can/can_esp32_twai.c:
```diff
@@ -18,12 +19,42 @@
 	const clock_control_subsys_t clock_subsys;
 	int irq_source;
+	/* 32-bit variant of output clock divider register required for non-ESP32 MCUs */
+	uint32_t cdr32;
```
I know, this *could* also be wrapped in an `#ifndef CONFIG_SOC_ESP32`, but I decided against it because it would make the macros at the bottom of the file much harder to read. 4 bytes of wasted RAM should be acceptable.

martinjaeger replied:
If you have a good idea how the `.cdr32 = CAN_ESP32_TWAI_DT_CDR_INST_GET(inst),` assignment can be excluded without duplicating the entire macro, let me know.

henrikbrixandersen replied:
Take a look at how the `FLEXCAN_PINCTRL_INIT` macro is for conditional config struct member assignment here: https://github.com/zephyrproject-rtos/zephyr/blob/147d9cb1272b39e119025953acdb3d19ba52e45e/drivers/can/can_mcux_flexcan.c#L953

henrikbrixandersen commented on drivers/can/can_esp32_twai.c:
```diff
@@ -113,14 +224,40 @@ const struct can_driver_api can_esp32_twai_driver_api = {
 };
 
-#define CAN_ESP32_TWAI_DT_CDR_INST_GET(inst)                                                       \
+#define CAN_SJA1000_DT_CDR_INST_GET(inst)                                                          \
```
This is still ESP32 TWAI binding specific and should be named as such.

henrikbrixandersen replied:
That'll work too. And yes, please leave the SJA1000 definitions in place for other frontends to use.

martinjaeger committed:
tests: drivers: can: api: add twai build-test for esp32c3

The TWAI peripheral needs a connected CAN transceiver also for testing
in loopback mode.

As there is no board available with this configuration, at least make
sure the CAN API test is built in CI for esp32c3_devkitm.

Signed-off-by: Martin Jäger <martin@libre.solar>

henrikbrixandersen commented:
Thanks @martinjaeger :)

sylvioalves reviewed (commented):
@martinjaeger, nice work, thanks!
